extract-requires.py

```diff
@@ -89,23 +86,28 @@ def get_build_backend_hook_caller(pyproject_toml):
             pyproject_toml = tomli.loads(f.read())
     hook_caller = get_build_backend_hook_caller(pyproject_toml)
 
-    requires = []
+    requires = set()
     if not (args.build_system or args.build_backend):
-        requires.extend(pyproject_toml.get('project', {}).get('dependencies', []))
+        requires.update(pyproject_toml.get('project', {}).get('dependencies', []))
 
         metadata_path = hook_caller.prepare_metadata_for_build_wheel("./")
 
         with open(os.path.join(metadata_path, "METADATA"), "r") as f:
             parsed = metadata.Metadata.from_email(f.read(), validate=False)
-            for r in (parsed.requires_dist or []):
-                if (not r.marker) or r.marker.evaluate(original_marker_env):
-                    requires.append(str(r))
-                else:
-                    print(f'ignoring {r} because marker evaluates false with context {original_marker_env}', file=sys.stderr)
+            for extra in original_requirement.extras if original_requirement.extras else [""]:
```

Comment:
With a requirement expression like `package[foo,bar]` the `extras` can be a list. Did you try just passing the right list to `r.marker.evaluate()`? I wouldn't be surprised if `extra = foo` is evaluated as `foo in ['foo', 'bar']` internally.

Reply:
Don't see any allowance for environment to be anything but a dict of strings

https://github.com/pypa/packaging/blob/e0dda88874e73cd484b9e8464c5921a903db3cf0/src/packaging/markers.py#L212-L222

Reply:
Huh. OK.